util/econftool.c

```diff
@@ -0,0 +1,557 @@
+ *
+ * @param path The path to be changed
+ */
+static void change_root_dir(char path[PATH_MAX])
+{
+    if (getenv("ECONFTOOL_ROOT") != NULL) {
+        if (strlen(path) + strlen(getenv("ECONFTOOL_ROOT")) >= PATH_MAX) {
+            fprintf(stderr, "ECONFTOOL_ROOT + path is too long\n");
+            exit(EXIT_FAILURE);
+        }
+
+        char *tmp = strdup(path);
+
+        strcpy(path, getenv("ECONFTOOL_ROOT"));
+        strcat(path, tmp);
+
+        free(tmp);
+    }
+}
+
+/**
+ * @brief This function is used as a callback for nftw to delete
+ *        all files nftw finds.
+ */
+static int nftw_remove(const char *path, const struct stat *sb, int flag, struct FTW *ftwbuf)
+{
+    return remove(path);
+}
+
+/**
+ * @brief Generate a tmpfile using mkstemp() and write its name into tmp_name
+ */
+static int generate_tmp_file(char *tmp_name)
+{
+    int filedes;
+    char name_template[21] = "/tmp/econftool-XXXXXX";
+    filedes = mkstemp(name_template);
+    if (filedes == -1) {
+        perror("mkstemp() failed");
+        return -1;
+    }
+
+    char fdpath[PATH_MAX];
+    char tmp_filename[PATH_MAX];
+    snprintf(fdpath, PATH_MAX, "/proc/self/fd/%d", filedes);
+    ssize_t r = readlink(fdpath, tmp_filename, PATH_MAX);
+    tmp_filename[r] = '\0';
+    strcpy(tmp_name, tmp_filename);
+    return 0;
+}
+
+/**
+ * @brief This command will read all snippets for filename.conf
+ *        (econf_readDirs) and print all groups, keys and their
+ *        values as an application would see them.
+ */
+static int econf_show(void)
+{
+    econf_err econf_error;
+
+    econf_error = econf_readDirs(&key_file, usr_root_dir, root_dir, filename, suffix, "=", "#");
+    if (econf_error) {
+        fprintf(stderr, "%s\n", econf_errString(econf_error));
+        econf_free(key_file);
+        return -1;
+    }
+
+    char **groups = NULL;
+    char *value = NULL;
+    size_t groupCount = 0;
+
+    /* show groups, keys and their value */
+    econf_error = econf_getGroups(key_file, &groupCount, &groups);
+    if (econf_error) {
+        fprintf(stderr, "%s\n", econf_errString(econf_error));
+        econf_free(groups);
+        econf_free(key_file);
+        return -1;
+    }
+    for (size_t g = 0; g < groupCount; g++)
+    {
+        char **keys = NULL;
+        size_t key_count = 0;
+
+        econf_error = econf_getKeys(key_file, groups[g], &key_count, &keys);
+        if (econf_error)
+        {
+            fprintf(stderr, "%s\n", econf_errString(econf_error));
+            econf_free(groups);
+            econf_free(keys);
+            return -1;
+        }
+        printf("%s\n", groups[g]);
+        for (size_t k = 0; k < key_count; k++) {
+            econf_error = econf_getStringValue(key_file, groups[g], keys[k], &value);
+            if (econf_error) {
+                fprintf(stderr, "%s\n", econf_errString(econf_error));
+                econf_free(groups);
+                econf_free(keys);
+                free(value);
```

Comment:
Is it guaranteed that `econf_getStringValue` returns a newly allocated `value` even on error?
If not, this would free the `value` of the previous iteration, causing undefined behaviour.